File: scripts/diff-redline.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { diffLines, diffWordsWithSpace } = require('diff');
const JSZip = require('jszip');
const {
  extractDocx,
  normalizeXML,
  getAllFiles,
  shouldIgnoreFile,
  filesAreIdentical,
  isXMLFile,
} = require('./diff-utils');

/**
 * DOCX Redlining Script - Creates a new DOCX with tracked changes
 *
 * This script compares two DOCX files and generates a third DOCX that shows
 * all differences as tracked changes (redlines) while preserving original styles.
 *
 * Usage: node scripts/diff-redline.js <baseline.docx> <current.docx> <output.docx>
 */

// Generate a random RSID (8-digit hex)
function generateRSID() {
  return Math.floor(Math.random() * 0xFFFFFFFF).toString(16).toUpperCase().padStart(8, '0');
}

// Get current date in Word format (YYYY-MM-DDTHH:MM:SSZ)
function getWordDate() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/**
 * Parse XML and extract text runs with their formatting
 * Returns array of { text, rPr, rsid } objects
 */
function parseTextRuns(xmlContent) {
  const runs = [];

  // Match <w:r> elements with their properties
  const runRegex = /<w:r(?: [^>]*)?>[\s\S]*?<\/w:r>/g;
  const matches = xmlContent.match(runRegex);

  if (!matches) return runs;

  for (const runXml of matches) {
    // Extract w:rPr (run properties) if present
    const rPrMatch = runXml.match(/<w:rPr>([\s\S]*?)<\/w:rPr>/);
    const rPr = rPrMatch ? rPrMatch[0] : '';

    // Extract RSID from w:r tag
    const rsidMatch = runXml.match(/w:rsidR="([^"]+)"/);
    const rsid = rsidMatch ? rsidMatch[1] : '';

    // Extract text from <w:t> elements
    const textMatches = runXml.match(/<w:t[^>]*>([\s\S]*?)<\/w:t>/g);
    if (textMatches) {
      const text = textMatches
        .map(t => t.replace(/<w:t[^>]*>|<\/w:t>/g, ''))
        .join('');

      runs.push({ text, rPr, rsid, fullXml: runXml });
    }
  }

  return runs;
}

/**
 * Parse paragraphs with their properties and runs
 */
function parseParagraphs(xmlContent) {
  const paragraphs = [];

  // Match <w:p> elements
  const paraRegex = /<w:p(?: [^>]*)?>[\s\S]*?<\/w:p>/g;
  const matches = xmlContent.match(paraRegex);

  if (!matches) return paragraphs;

  for (const paraXml of matches) {
    // Extract w:pPr (paragraph properties)
    const pPrMatch = paraXml.match(/<w:pPr>([\s\S]*?)<\/w:pPr>/);
    const pPr = pPrMatch ? pPrMatch[0] : '';

    // Extract runs
    const runs = parseTextRuns(paraXml);

    // Get full text
    const text = runs.map(r => r.text).join('');

    paragraphs.push({
      pPr,
      runs,
      text,
      fullXml: paraXml,
    });
  }

  return paragraphs;
}

/**
 * Create tracked change XML for a deletion
 */
function createDeletion(text, rPr, rsid, author = 'Claude', date = getWordDate()) {
  const delRsid = generateRSID();
  const xmlSpace = text.match(/^\s|\s$/) ? ' xml:space="preserve"' : '';

  return `<w:del w:id="${delRsid}" w:author="${author}" w:date="${date}">` +
    `<w:r${rsid ? ` w:rsidDel="${delRsid}"` : ''}>` +
    (rPr || '') +
    `<w:delText${xmlSpace}>${escapeXml(text)}</w:delText>` +
    `</w:r>` +
    `</w:del>`;
}

/**
 * Create tracked change XML for an insertion
 */
function createInsertion(text, rPr, rsid, author = 'Claude', date = getWordDate()) {
  const insRsid = generateRSID();
  const xmlSpace = text.match(/^\s|\s$/) ? ' xml:space="preserve"' : '';

  return `<w:ins w:id="${insRsid}" w:author="${author}" w:date="${date}">` +
    `<w:r${rsid ? ` w:rsidR="${insRsid}"` : ''}>` +
    (rPr || '') +
    `<w:t${xmlSpace}>${escapeXml(text)}</w:t>` +
    `</w:r>` +
    `</w:ins>`;
}

/**
 * Escape XML special characters
 */
function escapeXml(text) {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

/**
 * Create a redlined paragraph by comparing baseline and current
 */
function createRedlinedParagraph(baselinePara, currentPara) {
  if (!baselinePara && !currentPara) return '';

  // If paragraph is new (only in current)
  if (!baselinePara && currentPara) {
    const runs = currentPara.runs.map(run =>
      createInsertion(run.text, run.rPr, run.rsid)
    ).join('');
    return `<w:p>${currentPara.pPr}${runs}</w:p>`;
  }

  // If paragraph is deleted (only in baseline)
  if (baselinePara && !currentPara) {
    const runs = baselinePara.runs.map(run =>
      createDeletion(run.text, run.rPr, run.rsid)
    ).join('');
    return `<w:p>${baselinePara.pPr}${runs}</w:p>`;
  }

  // Both exist - compare text runs
  const baselineText = baselinePara.text;
  const currentText = currentPara.text;

  // If identical, use baseline as-is
  if (baselineText === currentText) {
    return baselinePara.fullXml;
  }

  // Perform word-level diff
  const diff = diffWordsWithSpace(baselineText, currentText);

  // Build redlined runs
  const redlinedRuns = [];
  let baselineRunIndex = 0;
  let baselineCharOffset = 0;
  let currentRunIndex = 0;
  let currentCharOffset = 0;

  for (const part of diff) {
    const text = part.value;

    if (!part.added && !part.removed) {
      // Unchanged text - use baseline runs with original formatting
      let remainingText = text;

      while (remainingText.length > 0 && baselineRunIndex < baselinePara.runs.length) {
        const run = baselinePara.runs[baselineRunIndex];
        const runText = run.text.substring(baselineCharOffset);

        if (runText.length <= remainingText.length) {
          // Use entire remaining run
          const rsidAttr = run.rsid ? ` w:rsidR="${run.rsid}"` : '';
          const xmlSpace = runText.match(/^\s|\s$/) ? ' xml:space="preserve"' : '';
          redlinedRuns.push(
            `<w:r${rsidAttr}>${run.rPr}<w:t${xmlSpace}>${escapeXml(runText)}</w:t></w:r>`
          );
          remainingText = remainingText.substring(runText.length);
          baselineRunIndex++;
          baselineCharOffset = 0;
        } else {
          // Use part of run
          const partText = runText.substring(0, remainingText.length);
          const rsidAttr = run.rsid ? ` w:rsidR="${run.rsid}"` : '';
          const xmlSpace = partText.match(/^\s|\s$/) ? ' xml:space="preserve"' : '';
          redlinedRuns.push(
            `<w:r${rsidAttr}>${run.rPr}<w:t${xmlSpace}>${escapeXml(partText)}</w:t></w:r>`
          );
          baselineCharOffset += remainingText.length;
          remainingText = '';
        }
      }

      currentCharOffset += text.length;
      if (currentRunIndex < currentPara.runs.length) {
        const currentRun = currentPara.runs[currentRunIndex];
        if (currentCharOffset >= currentRun.text.length) {
          currentRunIndex++;
          currentCharOffset = 0;
        }
      }
    } else if (part.removed) {
      // Deleted text - preserve baseline formatting
      let remainingText = text;

      while (remainingText.length > 0 && baselineRunIndex < baselinePara.runs.length) {
        const run = baselinePara.runs[baselineRunIndex];
        const runText = run.text.substring(baselineCharOffset);

        if (runText.length <= remainingText.length) {
          redlinedRuns.push(createDeletion(runText, run.rPr, run.rsid));
          remainingText = remainingText.substring(runText.length);
          baselineRunIndex++;
          baselineCharOffset = 0;
        } else {
          const partText = runText.substring(0, remainingText.length);
          redlinedRuns.push(createDeletion(partText, run.rPr, run.rsid));
          baselineCharOffset += remainingText.length;
          remainingText = '';
        }
      }
    } else if (part.added) {
      // Inserted text - use current formatting
      let remainingText = text;

      while (remainingText.length > 0 && currentRunIndex < currentPara.runs.length) {
        const run = currentPara.runs[currentRunIndex];
        const runText = run.text.substring(currentCharOffset);

        if (runText.length <= remainingText.length) {
          redlinedRuns.push(createInsertion(runText, run.rPr, run.rsid));
          remainingText = remainingText.substring(runText.length);
          currentRunIndex++;
          currentCharOffset = 0;
        } else {
          const partText = runText.substring(0, remainingText.length);
          redlinedRuns.push(createInsertion(partText, run.rPr, run.rsid));
          currentCharOffset += remainingText.length;
          remainingText = '';
        }
      }
    }
  }

  // Use baseline paragraph properties (preserve original structure)
  return `<w:p>${baselinePara.pPr}${redlinedRuns.join('')}</w:p>`;
}

/**
 * Generate redlined document.xml
 */
function generateRedlinedDocument(baselineXml, currentXml) {
  // Parse both documents
  const baselineParagraphs = parseParagraphs(baselineXml);
  const currentParagraphs = parseParagraphs(currentXml);

  // Extract document structure (before <w:body> and after </w:body>)
  const bodyStartMatch = baselineXml.match(/([\s\S]*?<w:body>)/);
  const bodyEndMatch = baselineXml.match(/(<\/w:body>[\s\S]*)/);

  const documentPrefix = bodyStartMatch ? bodyStartMatch[1] : '';
  const documentSuffix = bodyEndMatch ? bodyEndMatch[1] : '';

  // Perform paragraph-level diff
  const baselineTexts = baselineParagraphs.map(p => p.text);
  const currentTexts = currentParagraphs.map(p => p.text);

  const paraDiff = diffLines(baselineTexts.join('\n'), currentTexts.join('\n'));

  const redlinedParagraphs = [];
  let baselineIndex = 0;
  let currentIndex = 0;

  for (const part of paraDiff) {
    const lineCount = part.value.split('\n').length - 1;

    if (!part.added && !part.removed) {
      // Unchanged paragraphs - compare each one
      for (let i = 0; i < lineCount; i++) {
        const redlined = createRedlinedParagraph(
          baselineParagraphs[baselineIndex],
          currentParagraphs[currentIndex]
        );
        redlinedParagraphs.push(redlined);
        baselineIndex++;
        currentIndex++;
      }
    } else if (part.removed) {
      // Deleted paragraphs
      for (let i = 0; i < lineCount; i++) {
        const redlined = createRedlinedParagraph(
          baselineParagraphs[baselineIndex],
          null
        );
        redlinedParagraphs.push(redlined);
        baselineIndex++;
      }
    } else if (part.added) {
      // Inserted paragraphs
      for (let i = 0; i < lineCount; i++) {
        const redlined = createRedlinedParagraph(
          null,
          currentParagraphs[currentIndex]
        );
        redlinedParagraphs.push(redlined);
        currentIndex++;
      }
    }
  }

  return documentPrefix + redlinedParagraphs.join('') + documentSuffix;
}

/**
 * Enable track changes in settings.xml
 */
function enableTrackChanges(settingsXml) {
  // Add trackRevisions element if not present
  if (!settingsXml.includes('<w:trackRevisions')) {
    // Add after <w:proofState> or before </w:settings>
    const proofStateMatch = settingsXml.match(/(<w:proofState[^>]*\/>)/);
    if (proofStateMatch) {
      return settingsXml.replace(
        proofStateMatch[0],
        `${proofStateMatch[0]}<w:trackRevisions/>`
      );
    } else {
      return settingsXml.replace('</w:settings>', '<w:trackRevisions/></w:settings>');
    }
  }
  return settingsXml;
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length < 3) {
    console.error('Usage: node scripts/diff-redline.js <baseline.docx> <current.docx> <output.docx>');
    process.exit(1);
  }

  const baselinePath = args[0];
  const currentPath = args[1];
  const outputPath = args[2];

  if (!fs.existsSync(baselinePath)) {
    console.error(`Baseline file not found: ${baselinePath}`);
    process.exit(1);
  }

  if (!fs.existsSync(currentPath)) {
    console.error(`Current file not found: ${currentPath}`);
    process.exit(1);
  }

  console.log('📊 DOCX Redlining\n');
  console.log(`Baseline: ${baselinePath}`);
  console.log(`Current:  ${currentPath}`);
  console.log(`Output:   ${outputPath}\n`);

  // Extract both DOCX files
  const tempDir = path.join(__dirname, '..', '.tmp-redline');
  const baselineDir = path.join(tempDir, 'baseline');
  const currentDir = path.join(tempDir, 'current');

  console.log('Extracting DOCX files...');
  await extractDocx(baselinePath, baselineDir);
  await extractDocx(currentPath, currentDir);
  console.log('✓ Extraction complete\n');

  // Read document.xml from both
  const baselineDocPath = path.join(baselineDir, 'word', 'document.xml');
  const currentDocPath = path.join(currentDir, 'word', 'document.xml');

  const baselineXml = fs.readFileSync(baselineDocPath, 'utf8');
  const currentXml = fs.readFileSync(currentDocPath, 'utf8');

  console.log('Generating redlined document...');
  const redlinedXml = generateRedlinedDocument(baselineXml, currentXml);

  // Update settings.xml to enable track changes
  const settingsPath = path.join(baselineDir, 'word', 'settings.xml');
  if (fs.existsSync(settingsPath)) {
    const settingsXml = fs.readFileSync(settingsPath, 'utf8');
    const updatedSettings = enableTrackChanges(settingsXml);
    fs.writeFileSync(settingsPath, updatedSettings);
  }

  // Write redlined document.xml
  fs.writeFileSync(baselineDocPath, redlinedXml);

  console.log('✓ Redlining complete\n');

  // Create output DOCX from modified baseline
  console.log('Creating output DOCX...');
  const zip = new JSZip();

  // Read all files from baseline directory
  const allFiles = getAllFiles(baselineDir);

  for (const file of allFiles) {
    const filePath = path.join(baselineDir, file);
    const content = fs.readFileSync(filePath);
    zip.file(file, content);
  }

  const zipBuffer = await zip.generateAsync({
    type: 'nodebuffer',
    compression: 'DEFLATE',
    compressionOptions: { level: 9 }
  });

  fs.writeFileSync(outputPath, zipBuffer);
  console.log('✓ Output DOCX created\n');

  // Cleanup temp directory
  fs.rmSync(tempDir, { recursive: true, force: true });

  console.log(`✅ Redlined document saved to: ${outputPath}`);
  console.log('\nOpen in Microsoft Word to see tracked changes.');

  process.exit(0);
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: scripts/diff-redline.py
#!/usr/bin/env python3
"""DOCX Redlining Script - Creates tracked changes preserving styles.

Usage: python scripts/diff-redline.py <baseline.docx> <current.docx> <output.docx>

This script compares two DOCX files and generates a redlined version with tracked
changes, preserving all formatting and styles using the Document library.
"""

import sys
from difflib import SequenceMatcher
from pathlib import Path

# Add claude_office_skills to path
skills_path = Path(__file__).parent.parent / 'claude_office_skills'
sys.path.insert(0, str(skills_path))
sys.path.insert(0, str(skills_path / 'public' / 'docx'))

from public.docx.scripts.document import Document


def parse_runs_from_paragraph(para_elem):
    """Extract text runs with their formatting from a paragraph element.

    Returns list of dicts with: {text, rPr, rsid, elem}
    """
    runs = []
    for run in para_elem.getElementsByTagName('w:r'):
        # Extract text
        text_parts = [t_elem.firstChild.data for t_elem in run.getElementsByTagName('w:t') if t_elem.firstChild]
        text_parts.extend(
            dt_elem.firstChild.data for dt_elem in run.getElementsByTagName('w:delText') if dt_elem.firstChild
        )

        text = ''.join(text_parts)
        if not text:
            continue

        # Extract formatting properties
        rpr_elems = run.getElementsByTagName('w:rPr')
        rpr = rpr_elems[0].toxml() if rpr_elems else ''

        # Extract RSID
        rsid = run.getAttribute('w:rsidR') or run.getAttribute('w:rsidDel') or ''

        runs.append({'text': text, 'rPr': rpr, 'rsid': rsid, 'elem': run})

    return runs


def get_paragraph_properties(para_elem):
    """Extract paragraph properties (w:pPr) from paragraph element."""
    ppr_elems = para_elem.getElementsByTagName('w:pPr')
    return ppr_elems[0].toxml() if ppr_elems else ''


def diff_text_runs(baseline_runs, current_runs):
    """Compare two lists of text runs and generate diff operations.

    Returns list of operations: [(op, baseline_run, current_run), ...]
    op can be: 'equal', 'delete', 'insert', 'replace'
    """
    baseline_texts = [r['text'] for r in baseline_runs]
    current_texts = [r['text'] for r in current_runs]

    # Use SequenceMatcher for character-level diffing
    baseline_full = ''.join(baseline_texts)
    current_full = ''.join(current_texts)

    matcher = SequenceMatcher(None, baseline_full, current_full)
    operations = []

    for tag, i1, i2, j1, j2 in matcher.get_opcodes():
        baseline_chunk = baseline_full[i1:i2]
        current_chunk = current_full[j1:j2]

        if tag == 'equal':
            operations.append(('equal', baseline_chunk, None))
        elif tag == 'delete':
            operations.append(('delete', baseline_chunk, None))
        elif tag == 'insert':
            operations.append(('insert', None, current_chunk))
        elif tag == 'replace':
            operations.append(('replace', baseline_chunk, current_chunk))

    return operations, baseline_runs, current_runs


def apply_diff_to_paragraph(doc, para_elem, baseline_runs, current_runs, operations) -> None:
    """Apply diff operations to a paragraph using tracked changes.

    This preserves formatting from original runs while marking changes.
    """
    baseline_run_idx = 0
    baseline_char_offset = 0
    current_run_idx = 0
    current_char_offset = 0

    for op, baseline_chunk, current_chunk in operations:
        if op == 'equal':
            # Skip over unchanged text - don't modify it
            baseline_char_offset += len(baseline_chunk)
            current_char_offset += len(baseline_chunk)

        elif op == 'delete':
            # Find the run(s) containing deleted text and mark for deletion
            remaining = len(baseline_chunk)

            while remaining > 0 and baseline_run_idx < len(baseline_runs):
                run_info = baseline_runs[baseline_run_idx]
                run_elem = run_info['elem']
                run_text = run_info['text']

                # Check if run is already tracked
                parent = run_elem.parentNode
                if parent.tagName in {'w:ins', 'w:del'}:
                    # Skip already tracked changes
                    baseline_run_idx += 1
                    baseline_char_offset = 0
                    continue

                available = len(run_text) - baseline_char_offset

                if available <= remaining:
                    # Delete entire remaining part of run
                    if baseline_char_offset == 0:
                        # Delete whole run
                        doc['word/document.xml'].suggest_deletion(run_elem)
                    else:
                        # Need to split run - delete from offset onwards
                        # This is complex, for now mark entire run
                        doc['word/document.xml'].suggest_deletion(run_elem)

                    remaining -= available
                    baseline_run_idx += 1
                    baseline_char_offset = 0
                else:
                    # Delete part of run (complex - requires splitting)
                    # For now, mark entire run for deletion
                    doc['word/document.xml'].suggest_deletion(run_elem)
                    baseline_char_offset += remaining
                    remaining = 0

        elif op == 'insert':
            # Find appropriate insertion point and add tracked insertion
            remaining = len(current_chunk)

            while remaining > 0 and current_run_idx < len(current_runs):
                run_info = current_runs[current_run_idx]
                run_text = run_info['text']
                rpr = run_info['rPr']

                available = len(run_text) - current_char_offset

                if available <= remaining:
                    # Insert text with formatting
                    chunk_to_insert = run_text[current_char_offset:]
                    insert_xml = (
                        f'<w:ins><w:r>{rpr}<w:t xml:space="preserve">{escape_xml(chunk_to_insert)}</w:t></w:r></w:ins>'
                    )

                    # Find insertion point in baseline
                    if baseline_run_idx < len(baseline_runs):
                        anchor = baseline_runs[baseline_run_idx]['elem']
                        doc['word/document.xml'].insert_before(anchor, insert_xml)
                    else:
                        # Insert at end of paragraph
                        doc['word/document.xml'].append_to(para_elem, insert_xml)

                    remaining -= available
                    current_run_idx += 1
                    current_char_offset = 0
                else:
                    # Insert part of text
                    chunk_to_insert = run_text[current_char_offset : current_char_offset + remaining]
                    insert_xml = (
                        f'<w:ins><w:r>{rpr}<w:t xml:space="preserve">{escape_xml(chunk_to_insert)}</w:t></w:r></w:ins>'
                    )

                    if baseline_run_idx < len(baseline_runs):
                        anchor = baseline_runs[baseline_run_idx]['elem']
                        doc['word/document.xml'].insert_before(anchor, insert_xml)
                    else:
                        doc['word/document.xml'].append_to(para_elem, insert_xml)

                    current_char_offset += remaining
                    remaining = 0


def escape_xml(text):
    """Escape special XML characters."""
    return (
        text.replace('&', '&amp;')
        .replace('<', '&lt;')
        .replace('>', '&gt;')
        .replace('"', '&quot;')
        .replace("'", '&apos;')
    )


def compare_paragraphs(doc, baseline_paras, current_paras) -> None:
    """Compare paragraphs and generate tracked changes."""
    # Simple paragraph-level matching by text content
    baseline_texts = []
    for para in baseline_paras:
        runs = parse_runs_from_paragraph(para)
        text = ''.join(r['text'] for r in runs)
        baseline_texts.append((para, text, runs))

    current_texts = []
    for para in current_paras:
        runs = parse_runs_from_paragraph(para)
        text = ''.join(r['text'] for r in runs)
        current_texts.append((para, text, runs))

    # Match paragraphs by similarity
    matcher = SequenceMatcher(None, [t[1] for t in baseline_texts], [t[1] for t in current_texts])

    for tag, i1, i2, j1, j2 in matcher.get_opcodes():
        if tag == 'equal':
            # Compare each matched paragraph for run-level changes
            for i, j in zip(range(i1, i2), range(j1, j2)):
                baseline_para, baseline_text, baseline_runs = baseline_texts[i]
                current_para, current_text, current_runs = current_texts[j]

                if baseline_text != current_text:
                    # Paragraph content changed - apply run-level diff
                    operations, baseline_runs, current_runs = diff_text_runs(baseline_runs, current_runs)
                    apply_diff_to_paragraph(doc, baseline_para, baseline_runs, current_runs, operations)

        elif tag == 'delete':
            # Deleted paragraphs
            for i in range(i1, i2):
                baseline_para = baseline_texts[i][0]
                doc['word/document.xml'].suggest_deletion(baseline_para)

        elif tag == 'insert':
            # Inserted paragraphs
            for j in range(j1, j2):
                current_para, current_text, current_runs = current_texts[j]
                ppr = get_paragraph_properties(current_para)

                # Build paragraph with tracked insertion
                runs_xml = [
                    f'<w:r>{run["rPr"]}<w:t xml:space="preserve">{escape_xml(run["text"])}</w:t></w:r>'
                    for run in current_runs
                ]

                para_xml = f'<w:p>{ppr}{"".join(runs_xml)}</w:p>'
                tracked_para = doc['word/document.xml'].suggest_paragraph(para_xml)

                # Find insertion point
                if i1 < len(baseline_texts):
                    anchor = baseline_texts[i1][0]
                    doc['word/document.xml'].insert_before(anchor, tracked_para)
                else:
                    # Insert at end
                    body = doc['word/document.xml'].get_node(tag='w:body')
                    doc['word/document.xml'].append_to(body, tracked_para)


def main() -> None:
    if len(sys.argv) < 4:
        sys.exit(1)

    baseline_docx = Path(sys.argv[1])
    current_docx = Path(sys.argv[2])
    output_docx = Path(sys.argv[3])

    if not baseline_docx.exists():
        sys.exit(1)

    if not current_docx.exists():
        sys.exit(1)

    doc = None
    current_doc = None
    try:
        # Initialize Document with baseline (makes a copy)
        doc = Document(baseline_docx, track_revisions=True)

        # Get paragraphs from both documents
        baseline_paras = doc['word/document.xml'].dom.getElementsByTagName('w:p')

        # Load current document for comparison
        current_doc = Document(current_docx, track_revisions=False)
        current_paras = current_doc['word/document.xml'].dom.getElementsByTagName('w:p')

        # Perform comparison and apply tracked changes
        compare_paragraphs(doc, list(baseline_paras), list(current_paras))

        # Save the redlined document directly to output
        doc.save(destination=output_docx)
    finally:
        if doc is not None:
            doc.close()
        if current_doc is not None:
            current_doc.close()


if __name__ == '__main__':
    main()
